File: gui/src/app/apps/canary/docker/docker.service.ts
import {Injectable} from "@angular/core";
import {HttpClient} from "@angular/common/http";

@Injectable()
export class DockerService {

  constructor(private http: HttpClient) {
  }

  loadList() {
    return this.http.get<DockerImageSummary[]>('/rest/canary/docker/images')
  }

}

export interface DockerImageSummary {
  identifier: string,
  tag: string,
  state?: DockerImageState
}

export interface DockerImageState {
  digest: string,
  tags: string[],
  recordedAt: string,
}

File: canary-service/src/main/kotlin/ch/awae/mycloud/service/canary/dockerhub/service/DockerService.kt
package ch.awae.mycloud.service.canary.dockerhub.service

import ch.awae.mycloud.service.canary.dockerhub.dto.*
import ch.awae.mycloud.service.canary.dockerhub.model.*
import jakarta.transaction.*
import org.springframework.stereotype.*
import org.springframework.util.comparator.Comparators
import java.time.LocalDateTime

@Transactional
@Service
class DockerService(
    val monitoredEntryRepository: MonitoredEntryRepository
) {

    fun listAll(): List<DockerImageSummary> {
        return monitoredEntryRepository.findAll()
            .sortedByDescending { it.currentState?.recordedAt ?: LocalDateTime.MAX }
            .map {
                DockerImageSummary(
                    identifier = it.webIdentifier,
                    tag = it.tag,
                    state = it.currentState?.let { s ->
                        DockerImageSummary.State(
                            digest = s.digest,
                            tags = s.tags.sorted(),
                            recordedAt = s.recordedAt.toString().replace("T", " "),
                        )
                    },
                )
            }
    }

}